Bug fixes related to solar diag package: COS(zenith angle), non-existent hydrometeors, shortwave irradiances in time series

twjuliano opened the pull request:
TYPE: bug fix, enhancement

KEYWORDS: solar diagnostic, shortwave irradiance, solar zenith, hydrometeor, cloud

SOURCE: Timothy W. Juliano, Ju-Hye Kim, Pedro A. Jimenez, Jared Lee, Thomas Brummet (NCAR/RAL)

DESCRIPTION OF CHANGES:
**Bug fixes**
Problem:
     1. Value of cosine solar zenith angle (coszen) can be slightly <-1 or slightly >1, which results in a nonphysical value for diagnosis of the solar zenith angle
     2. Calculating cloud parameters in solar diagnostic package when hydrometeor type is not present

Solution:
     1. Bound the values of coszen to be [-1, 1]
     2. Add conditional check in solar diagnostic module for presence of hydrometeor type before calculating cloud parameters. Also add check for summation of water path variables to ensure >=0.0 kg/m2.

**Enhancement**
We have added the Global Horizontal Irradiance (SWDOWN), Direct Normal Irradiance (SWDDNI), and Diffuse Horizontal Irradiance (SWDDIF), as well as the clear-sky Global Horizontal Irradiance (SWDOWNC) and clear-sky Direct Normal Irradiance (SWDDNIC) variables to the tslist when the solar diagnostic package is activated. Furthermore, we add these five variables as calculated by FARMS: SWDOWN2, SWDDNI2, SWDDIF2, SWDOWNC2, and SWDDNIC2.

All 10 variables are output regardless of model physics configuration. Variables SWDOWN, SWDDNI, and SWDDIF are calculated when any shortwave radiation scheme is active. If the RRTMG (ra_sw_physics=4) or RRTMG FAST (ra_sw_physics=24) shortwave radiation scheme or FARMS (swint_opt=2) is activated, then variables SWDOWNC and SWDDNIC are calculated [otherwise, set to MISSING (=-999.0)]. If FARMS (swint_opt=2) is activated, then variables SWDOWN2, SWDDNI2, SWDDIF2, SWDOWNC2, and SWDDNIC2 are calculated [otherwise, set to MISSING (=-999.0)].

LIST OF MODIFIED FILES:
     M Registry/Registry.EM_COMMON
     M Registry/registry.solar_fields
     M phys/module_diag_solar.F
     M phys/module_diagnostics_driver.F
     M phys/module_radiation_driver.F
     M run/README.tslist
     M share/wrf_timeseries.F

TESTS CONDUCTED: 
     1. We have run simulations with and without the fixes. Problem 1: When coszen is within the range [-1, 1], the model results are bit for bit identical before and after the change. Problem 2: The model output confirms that the cloud parameters are not calculated when the hydrometeor type is not present and also that the water path variables are positive. Enhancement: The tslist output is written as expected.
     2. Jenkins testing is all PASS.

RELEASE NOTE: Several fixes are introduced for the solar diagnostic package: 1) Correct solar zenith angle calculation; and 2) check presence of hydrometeor type before calculating cloud parameters. An enhancement to the package is also introduced: Adding shortwave irradiance variables to the tslist when activating the RRTMG or RRTMG FAST shortwave radiation scheme or FARMS.

davegill commented on phys/module_diag_solar.F:
```diff
@@ -39,6 +39,7 @@ SUBROUTINE solar_diag (rho, dz8w, ph, phb, cldfrac3d, coszen, swdnb, swdnt, qv,
 !----------------------------------------------------------------------
 
      USE module_model_constants, ONLY: G
+     USE module_state_description, ONLY: param_first_scalar, p_qc, p_qi, p_qs
```
@twjuliano @dudhia @weiwangncar 
Tim,
Would it be better to have these come through an argument list. That would fit better with an eventual paradigm we are heading towards.

twjuliano replied:
@davegill 
Dave, this has been fixed.

davegill commented on phys/module_radiation_driver.F:
```diff
@@ -1049,9 +1049,15 @@ SUBROUTINE radiation_driver (                                          &
-         call calc_coszen(ims,ime,jms,jme,its,ite,jts,jte,  &
-                          julian,xtime+radt*0.5,gmt, &
-                          declin,degrad,xlong,xlat,coszen,hrang)
+         if ( solar_opt_local == 1 ) then
```
@twjuliano 
Tim,
We probably need a short comment to explain why we have these two different times in otherwise identical subroutine calls based on a flag value.

twjuliano replied:
@davegill 
Dave, this has been added.

davegill commented:
@twjuliano 
Tim,
In the PR title, include "incorrect time" and "cloud parameters" for more specificity.

twjuliano commented:
@davegill @pedro-jm 
Dave,
I have addressed your comments above.

We have also added a fix for the coszen calculation so that it does not result in nonphysical values when the solar zenith angle is diagnosed. Moreover, shortwave irradiance variables from WRF-Solar are added to the tslist when the solar diagnostic package is activated. We forgot to include these previously, and they are important variables to output when using WRF-Solar.

Thanks,
Tim

davegill commented:
@weiwangncar @dudhia 
Folks,
Are we OK that there are bug fixes and new enhancements together in a single PR? I can go either way on this. Are there existing post-processors that would have trouble reading the newly manufactured files? Seems like it is just extra data that is tagged on the end of a formatted WRITE, so there should be no impact.

davegill commented:
@twjuliano 
> Furthermore, we add these five variables as calculated by FARMS if the scheme is activated (swint_opt=2). These variables from FARMS are: SWDOWN2, SWDDNI2, SWDDIF2, SWDOWNC2, and SWDDNIC2.

Tim,
So the additional variables are not output unless swint_opt=2? Has this been run without swint_opt=2 to verify the behavior is OK?

twjuliano commented:
@davegill 

> Tim,
> So the additional variables are not output unless swint_opt=2? Has this been run without swint_opt=2 to verify the behavior is OK?

Dave,
Apologies for the confusion; the additional variables are output regardless of namelist configuration. The values are set to MISSING (=-999.0) depending upon the namelist options. Therefore, no issues with tslist writing behavior. I've clarified this in the PR text.

weiwangncar commented:
I would prefer the bug fixes are separated from enhancement, if this is not too much to ask.. And that is definitely the thing to do in the future. If they are separated, it could mean the enhancement may not get into release branch until next year...

pedro-jm commented:
Adding the time series of the irradiances can be considered a bug fix on our end. The main motivation for developing the solar diagnostic package was to have the irradiance time series which was the last of our old WRF-Solar functionality to be ported. And somehow we forgot to add these vars! If we cannot do it now we'll wait obviously.

dudhia commented:
I have no problem with this as I think it only affects the time series.
Which modules are affected by this?

On Mon, Sep 14, 2020 at 4:17 PM pedro-jm <notifications@github.com> wrote:

> Adding the time series of the irradiances can be considered a bug fix on
> our end. The main motivation for developing the solar diagnostic package
> was to have the irradiance time series which was the last of our old
> WRF-Solar functionality to be ported. And somehow we forgot to add these
> vars! If we cannot do it now we'll wait obviously.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1284#issuecomment-692343112>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77HNSUF2DYAAZRA2J3DSF2JAXANCNFSM4RH5U5LA>
> .
>

twjuliano commented:
@dudhia 

> I have no problem with this as I think it only affects the time series. Which modules are affected by this?

That is correct. Aside from Registry files and README.tslist, only share/wrf_timeseries.F is impacted by this.

vikramjakhr commented:
Reopening for testing

weiwangncar commented:
@twjuliano Can you confirm that this PR does not impact WRF model results apart from FARMS output and time series?

twjuliano commented:
> @twjuliano Can you confirm that this PR does not impact WRF model results apart from FARMS output and time series?

@weiwangncar If solar_diagnostics=1, then this PR changes the model results because the coszen calculation is different, which ultimately impacts the radiation calculations. However, if solar_diagnostics=0, then this this PR does not impact WRF model results with the exception of instances when coszen <-1 or > 1. In my tests, the model results before and after the PR using solar_diagnostics=0 were identical using diffwrf, so @pedro-jm will have to comment on how often this coszen bounding may occur, as I have not experienced this issue. Thanks.

weiwangncar commented:
@twjuliano Can you explain the last change?

twjuliano commented:
> @twjuliano Can you explain the last change?

@weiwangncar After discussing with @pedro-jm we have decided that my modification for including a different time in the subroutine call for calc_coszen is not correct. I initially made this modification because the output coszen value is not correct when the solar diagnostic package is activated. However, I did not realize that it would change the code in an undesirable way. Therefore, for this PR, we will leave this issue untouched. We will think about the best approach to fix it and leave it for a future PR. Apologies for the confusion.

Now, with the latest commit, I can confirm that this PR does not impact WRF model results (i.e., bit for bit reproducibility between simulations using the old and new code).

weiwangncar commented:
@twjuliano Thanks. @dudhia Can you comment on the issue of coszen value falling out of the range [-1,1]?

dudhia commented:
I think that change was previously not correct because it would change the
results with the diagnostics on
due to the change in effective xtime.

On Thu, Oct 8, 2020 at 4:48 PM weiwangncar <notifications@github.com> wrote:

> @twjuliano <https://github.com/twjuliano> Can you explain the last change?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1284#issuecomment-705862763>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77A77HITCQ36HH6XECTSJY6VRANCNFSM4RH5U5LA>
> .
>

dudhia commented:
Regarding the limiting. I am not sure how often it happens or how far above 1 it gets in magnitude. Would be good to know. Adding this limit is a bandaid, but I think the diagnostics need it within this range to work, so it is valid.

twjuliano commented:
> I think that change was previously not correct because it would change the results with the diagnostics on due to the change in effective xtime.
> […](#)
> On Thu, Oct 8, 2020 at 4:48 PM weiwangncar ***@***.***> wrote: @twjuliano <https://github.com/twjuliano> Can you explain the last change? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#1284 (comment)](https://github.com/wrf-model/WRF/pull/1284#issuecomment-705862763)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AEIZ77A77HITCQ36HH6XECTSJY6VRANCNFSM4RH5U5LA> .

@dudhia Yes, that is correct.


> Regarding the limiting. I am not sure how often it happens or how far above 1 it gets in magnitude. Would be good to know. Adding this limit is a bandaid, but I think the diagnostics need it within this range to work, so it is valid.

@dudhia I will defer to @pedro-jm regarding how often it happens, as he introduced this bug fix. Thanks.

pedro-jm commented:
I do not think this happens very often. I took a look at one point and coszen was slightly lower than -1 or slightly higher than 1 in one time step only. It was something like a round off error. Although small, the solar diagnostic package needs it because the solar zenit angle is calculated as the inverse function of the cosine (arc cos). If cosezen does not belong to [-1, 1] then we have a bug.
Pedro.

dudhia commented:
Seems reasonable to have. It is the only change outside the diagnostics.

On Fri, Oct 9, 2020 at 3:26 PM pedro-jm <notifications@github.com> wrote:

> I do not think this happens very often. I took a look at one point and
> coszen was slightly lower than -1 or slightly higher than 1 in one time
> step only. It was something like a round off error. Although small, the
> solar diagnostic package needs it because the solar zenit angle is
> calculated as the inverse function of the cosine (arc cos). If cosezen does
> not belong to [-1, 1] then we have a bug.
> Pedro.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1284#issuecomment-706409364>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77BFZRX3F2RF3VKDHJ3SJ55YRANCNFSM4RH5U5LA>
> .
>

weiwangncar commented:
@dudhia Should we move this change to a separate PR? This will have an impact, however small, to the bit wise results.

dudhia commented:
I think it’s ok. It is needed for the new code.
Jimy


On Fri, Oct 9, 2020 at 4:25 PM weiwangncar <notifications@github.com> wrote:

> @dudhia <https://github.com/dudhia> Should we move this change to a
> separate PR? This will have an impact, however small, to the bit wise
> results.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1284#issuecomment-706423535>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77ANP2FZXXIZBHEI7X3SJ6EVZANCNFSM4RH5U5LA>
> .
>

dudhia commented:
When the diagnostics are set as MISSING how do they appear in the output versus before the fix?

weiwangncar commented:
@twjuliano Would you like to comment on Jimy's last post?

davegill commented:
@weiwangncar @dudhia 
Folks,
So the status of this PR: 
1. The "enhancement" portion (adding in the time series variables) really is also part of the bug fix, as the time series code was accidentally left out of the original PR. I am OK with that explanation.
2. I agree that the desired bounding of the COS to be between [-1,1] should be included. I am not concerned about bit-wise differences that occur when necessary fixes are applied.
3. The tests for cloud, ice, and snow account for unavailable fields in the computations.

From a position of evaluating this PR mechanically, I am OK with this code.

davegill commented on phys/module_diag_solar.F:
```diff
@@ -252,78 +255,99 @@ SUBROUTINE solar_diag (rho, dz8w, ph, phb, cldfrac3d, coszen, swdnb, swdnt, qv,
              end if
 
-             if (f_qc .or. f_qi .or. f_qs) then
+             if ( (f_qc .or. f_qi .or. f_qs) .and. & 
+                  (p_qc > param_first_scalar .or. &
+                   p_qi > param_first_scalar .or. &
+                   p_qs > param_first_scalar) ) then
```
Here we check that if any of Qc, Qi, or Qs are available

davegill commented on phys/module_diag_solar.F:
```diff
@@ -252,78 +255,99 @@ SUBROUTINE solar_diag (rho, dz8w, ph, phb, cldfrac3d, coszen, swdnb, swdnt, qv,
-               end do
+                 do while ( (wc < WC_MIN) .and. (k < kte) )
+                   k = k + 1
+                   wc = qc(i, k, j) + qi(i, k, j) + qs(i, k, j)
```
... but here we assume that they all are available
wc = qc(i, k, j) + qi(i, k, j) + qs(i, k, j)

Has this been tested with a simple MP scheme, such as WSM3?

twjuliano replied:
@davegill  Thanks for pointing this out. You are right -- this does appear to be an inconsistency. I just tested with WSM3, and variable wc is calculated correctly (i.e., variables qi and qs are equal to 0.0 such that wc = qc). Results were the same when compiling in debug mode. I can fix the code to be more explicit about which hydrometeor classes are considered if you would like. Thanks.

weiwangncar replied:
@davegill @twjuliano Actually when you use WSM3, qi and qs are not in the memory. So this code is probably not safe.

davegill replied:
@twjuliano 
Tim,
We definitely need some IF tests here. Just above, you use f_qc and f_qi (and p_qc and p_qi). Those existence tests should be part of this assignment.

We are hoping to push the release-v4.2.2 out next week. No pressure.

davegill replied:
@weiwangncar 
Wei,
We need to have some logic in place to safeguard against use of variables that are not allocated. Even if things appear OK, this is still a concern.

davegill replied:
@twjuliano 
Tim, 
Wow, that extra logic for qi and qs does really muddy up the ease of understanding! 

This change looks to be exactly what we were asking for. 

In the comments below, would you state that you have run one of your test cases with this new logic and that the results are still OK.

twjuliano replied:
@davegill 
Great. I've edited my latest comment at the bottom.

davegill commented:
jenkins
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 19d71b854a5eabae8a7bc82fbcf48533628b05f0, requested by: twjuliano for PR: https://github.com/wrf-model/WRF/pull/1284. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 166           164        0
    Number of Comparisons  : 105           104        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

twjuliano commented:
> When the diagnostics are set as MISSING how do they appear in the output versus before the fix?

> @twjuliano Would you like to comment on Jimy's last post?

@dudhia @weiwangncar  Before the fix, the diagnostics would have value '0' and now they show '-999' (i.e., MISSING).

weiwangncar commented:
@davegill Can you comment on my last comment?

davegill commented:
@weiwangncar 
Thanks for the reminder. I commented in the block of code for module_solar_diag.F, lines 278-280.

twjuliano commented:
@weiwangncar @davegill 
Wei and Dave-- Thank you for your help to make this code up to standards. Hopefully my latest commit resolves any issues.

davegill reviewed (approved):
Approved
